File: Top-Interview-Questions/Strings/242-Valid-Anagram.cpp
// Time Complexity: O(n) length of string
// Space complexity: O(1) only 26 characters * 2
// Runtime: 16 ms, faster than 72.28%
// Memory Usage: 7.5 MB, less than 100.00%

// Best Case: string is null or one
// Worst Case: string is longer than one character
class Solution {
public:
    bool isAnagram(string s, string t) {
        
        vector<int> a(26,0), b(26,0);
        
        if(s.length() != t.length())
            return false;
        
        for(int i = 0; i < s.length(); i++){
            a[s[i] - 'a']++;
            b[t[i] - 'a']++;
        }
                
        return a == b;
    }
};

File: Top-Interview-Questions/Strings/242-valid-anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 * Runtime: 68 ms, faster than 81.22% of JavaScript online submissions for Valid Anagram.
 * Memory Usage: 39.3 MB, less than 18.37% of JavaScript online submissions for Valid Anagram.
 * TC: O(s.length + t.length) which is linear
 * SC: O(s.length)
 */
var isAnagram = function(s, t) {
    if (s.length !== t.length) {
        return false;
    }

    const map = {};

    for (const char of s) {
        map[char] = map[char] ? ++map[char] : 1;
    }

    for (const char of t) {
        if (!map[char]) {
            return false;
        } else if (map[char] === 0) {
            return false;
        } else {
            map[char]--;
        }
    }

    return true;
};